src/hooks/useQuery.ts

```diff
@@ -0,0 +1,87 @@
+//libs
+import { useCallback, useEffect, useState } from "react";
+
+//constants
+import { STATUS } from "../Constants/constants";
+
+
+type statusProps = {
+  status: string;
+  data: any; // Generic
+  error: Error | undefined;
+};
+
+export const useQuery =(
+  asyncFunc: (asyncFuncParams: string) => Promise<Response>,
+  asyncFuncParams: string,
+  skip: boolean
+) => {
+  const [state, setState] = useState<statusProps>({
+    status: STATUS.IDLE,
+    data: undefined,
+    error: undefined,
+  });
+
+  useEffect(() => {
+    //to ignore network request
+    let ignore = false;
+    if (skip || ignore) return;
+    setState((state) => {
+      return {
+        ...state,
+        status: STATUS.LOADING,
+      };
+    });
+    asyncFunc(asyncFuncParams)
+      .then((response) => {
+        if (!response.ok) {
+          setState((state) => {
+            return {
+              ...state,
+              error: new Error("Error from backend"),
+            };
+          });
+          throw new Error();
```

Comment:
No need to setState here. Just throw the error. Catch block should handle the state update.